miniapps/meshing/minimal-surface.cpp

```diff
@@ -0,0 +1,1326 @@
+   const int Q1D = IntRules.Get(Geometry::SEGMENT, ir.GetOrder()).GetNPoints();
+   MFEM_VERIFY(ND == D1D*D1D, "");
+   MFEM_VERIFY(NQ == Q1D*Q1D, "");
+
+   Vector Eu(ND*NE), grad_u(DIM*NQ*NE), sum(NE*NQ), one(NE*NQ);
+   qi->SetOutputLayout(QVectorLayout::byVDIM);
+   const ElementDofOrdering e_ordering = ElementDofOrdering::LEXICOGRAPHIC;
+   const Operator *G(fes.GetElementRestriction(e_ordering));
+   G->Mult(u, Eu);
+   qi->Derivatives(Eu, grad_u);
+
+   auto W = Reshape(ir.GetWeights().Read(), Q1D, Q1D);
+   auto J = Reshape(geom->J.Read(), Q1D, Q1D, DIM, DIM, NE);
+   auto detJ = Reshape(geom->detJ.Read(), Q1D, Q1D, NE);
+   auto grdU = Reshape(grad_u.Read(), DIM, Q1D, Q1D, NE);
+   auto S = Reshape(sum.Write(), Q1D, Q1D, NE);
+
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, 1,
+   {
+      MFEM_FOREACH_THREAD(qy,y,Q1D)
+      {
+         MFEM_FOREACH_THREAD(qx,x,Q1D)
+         {
+            const double w = W(qx, qy);
+            const double J11 = J(qx, qy, 0, 0, e);
+            const double J12 = J(qx, qy, 1, 0, e);
+            const double J21 = J(qx, qy, 0, 1, e);
+            const double J22 = J(qx, qy, 1, 1, e);
+            const double det = detJ(qx, qy, e);
+            const double area = w * det;
+            const double gu0 = grdU(0, qx, qy, e);
+            const double gu1 = grdU(1, qx, qy, e);
+            const double tgu0 = (J22*gu0 - J12*gu1)/det;
+            const double tgu1 = (J11*gu1 - J21*gu0)/det;
+            const double ngu = tgu0*tgu0 + tgu1*tgu1;
+            const double s = (ker == AREA) ? sqrt(1.0 + ngu) :
+            (ker == NORM) ? ngu : 0.0;
+            S(qx, qy, e) = area * s;
+         }
+      }
+   });
+   one = 1.0;
+   return sum * one;
+}
+
+static int Problem1(Opt &opt)
+{
+   const int order = opt.order;
+   Mesh mesh(opt.nx, opt.ny, QUAD);
+   mesh.SetCurvature(opt.order, false, DIM, Ordering::byNODES);
+   for (int l = 0; l < opt.refine; l++) { mesh.UniformRefinement(); }
+   const H1_FECollection fec(order, DIM);
+   FiniteElementSpace fes(&mesh, &fec);
+   Array<int> ess_tdof_list;
+   if (mesh.bdr_attributes.Size())
+   {
+      Array<int> ess_bdr(mesh.bdr_attributes.Max());
+      ess_bdr = 1;
+      fes.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+   }
+   GridFunction uold(&fes), u(&fes), b(&fes);
+   FunctionCoefficient u0_fc(u0);
+   u.ProjectCoefficient(u0_fc);
+   if (opt.vis) { opt.vis = glvis.open(vishost, visport) == 0; }
+   if (opt.vis) { Surface::Visualize(opt, &mesh, GLVIZ_W, GLVIZ_H, &u); }
+   Vector B, X;
+   OperatorPtr A;
+   GridFunction eps(&fes);
+   for (int i = 0; i < opt.niters; i++)
+   {
+      b = 0.0;
+      uold = u;
+      BilinearForm a(&fes);
+      if (opt.pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+      const double q_uold = qf(order, AREA, mesh, fes, uold);
+      MFEM_VERIFY(std::fabs(q_uold) > EPS,"");
+      ConstantCoefficient q_uold_cc(1.0/sqrt(q_uold));
+      a.AddDomainIntegrator(new DiffusionIntegrator(q_uold_cc));
```

Comment:
Here and also in the parallel version the ```BilinearForm```, ```ConstantCoefficient```, and ```CGSolver``` objects could probably be hoisted out of the loop, so they aren't constantly created each iteration. However, this problem appears to run fairly fast, so I'm not too sure it's that big of a deal for this miniapp.

Reply:
Thanks for the review @rcarson3 , the `BilinearForm` is using the `DiffusionIntegrator` with the `ConstantCoefficient` built from `uold` so I kept it in for-loop body, but I moved the CG out of it as you suggested.

Reply:
`done`